pick_kernel/kernel.py

```diff
@@ -151,19 +149,60 @@ async def start_kernel(self, config=None):
             "-c",
             f"""the_config = '''{config}''';""",
         ]
+        self.log.info("launching child kernel with")
         self.log.info(km.kernel_cmd)
+
         await km.start_kernel()
 
         kernel = KernelProxy(manager=km, shell_upstream=self.shell_stream)
         self.iosub.connect(kernel.iopub_url)
 
-        # TODO: Make sure the kernel is really started
-        #       We can do that using kernel_info_requests as well as looking
-        #       at the kernel logs. We can use both of those to customize
-        #       the information we send back to the user.
+        # Make sure the kernel is really started. We do that by using
+        # kernel_info_requests here.
         #
-        # For now, we'll pretend the kernel takes 1.5 seconds to start
-        await asyncio.sleep(1.5)
+        # In the future we should incorporate kernel logs (output of the kernel process), then
+        # and send back all the information back to the user as display output.
+
+        # Create a temporary KernelClient for waiting for the kernel to start
+        kc = km.client()
+        kc.start_channels()
+
+        # Wait for kernel info reply on shell channel
+        while True:
+            self.log.debug("querying kernel info")
+            kc.kernel_info(reply=False)
+            try:
+                msg = await kc.shell_channel.get_msg(timeout=1)
+            except Empty:
+                pass
+            else:
+                if msg["msg_type"] == "kernel_info_reply":
+                    # Now we know the kernel is (mostly) ready.
+                    # However, most kernels are not quite ready at this point to send
+                    # on more execution.
+                    #
+                    # Do we wait for a further status: idle on iopub?
+                    # Wait for idle?
+                    # Wait for particular logs from the stdout of the kernel process?
```

Comment:
Hmm probably don't want to wait for logs... waiting for idle may be valid, but getting the kernel_info_reply does mean the process is alive. Might be ok to just let the downstream kernel continue processing messages from here as normal?

Reply:
The most common issue I ran into was that shell would be ready but iopub would not, and so execution messages would happen and we'd miss some batch of iopub. This was easiest to reproduce with a restart-and-run-all command from the Jupyter UI.